File: solutions/0278.First-Bad-Version/Solution.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        // using binary search
        int l = 1, r = n;
        int firstBadVer = Integer.MAX_VALUE; 
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (isBadVersion(m)) {
                if (firstBadVer > m) 
                    firstBadVer = m;
                r = m - 3;
            }
            else
                l = m + 1;
        }
        return firstBadVer;
    }
}

File: solutions/0278.First-Bad-Version/Solution.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
       var l = 1, r =  n;
        var firstBadVersion = n + 1;
        while (l <= r) {
            var m = l + Math.floor((r - l) / 2);
            if (isBadVersion(m)) {
                if (firstBadVersion > m)
                    firstBadVersion = m;
                r = m - 1;
            } else
                l = m + 1;
        }
        return firstBadVersion;
    };
};
